File: src/geometry/bounding.cpp
#include "geometry/bounding.hpp"
#include "geometry/triangle.hpp"

#include <glm/glm.hpp>

using namespace glm;
using namespace std;

namespace geometry {
Aabb find_bounding(const vector<Triangle>& triangles) {
  vec3 min, max;

  for (const Triangle& tri : triangles) {
    min = glm::min(min, tri.v0);
    min = glm::min(min, tri.v1);
    min = glm::min(min, tri.v2);

    max = glm::max(max, tri.v0);
    max = glm::max(max, tri.v1);
    max = glm::max(max, tri.v2);
  }

  vec3 half = (max - min) / 2.0f;
  return {min + half, half};
}
}

File: src/geometry/bounding.rs
use crate::geometry::aabb::Aabb;
use crate::geometry::triangle::Triangle;

pub fn bounding(triangles: &[Triangle]) -> Aabb {
    if triangles.is_empty() {
        return Aabb::empty()
    }
    let mut a = triangles[0].min();
    let mut b = triangles[0].max();
    for triangle in triangles {
        a = a.inf(&triangle.min());
        b = b.sup(&triangle.max());
    }
    Aabb::from_extents(&a, &b)
}

#[cfg(test)]
mod tests {
    use super::*;
    use nalgebra::vector;

    #[test]
    fn test_bounding() {
        let triangles = [
            Triangle{ v0: vector![1.0, 1.0, 0.0], v1: vector![1.0, 1.0, 1.0], v2: vector![0.0, 0.0, 0.0] },
            Triangle{ v0: vector![-1.0, -1.0, 0.0], v1: vector![-1.0, -1.0, -1.0], v2: vector![0.0, 0.0, 0.0] },
        ];

        let actual = bounding(&triangles);

        let expected = Aabb {
            center: vector![0.0, 0.0, 0.0],
            half_size: vector![1.0, 1.0, 1.0]
        };
        assert_eq!(actual, expected);
    }
}
